Content.Server/Antag/AntagSelectionSystem.cs

```diff
@@ -1,347 +1,451 @@
+        _inventory.SpawnItemsOnEntity(player, EntitySpawnCollection.GetSpawns(def.Equipment));
+
+        if (session != null)
+        {
+            var curMind = session.GetMind();
+            if (curMind == null)
+            {
+                curMind = _mind.CreateMind(session.UserId, Name(antagEnt.Value));
+                _mind.SetUserId(curMind.Value, session.UserId);
+            }
+
+            EntityManager.AddComponents(curMind.Value, def.MindComponents);
+            _mind.TransferTo(curMind.Value, antagEnt, ghostCheckOverride: true);
+            ent.Comp.SelectedMinds.Add((curMind.Value, Name(player)));
+        }
+
+        if (def.Briefing is { } briefing)
+        {
+            SendBriefing(session, Loc.GetString(briefing.Text), briefing.Color, briefing.Sound);
+        }
+
+        var afterEv = new AfterAntagEntitySelectedEvent(session, player, ent, def);
+        RaiseLocalEvent(ent, ref afterEv, true);
     }
-    #endregion
 
-    #region Briefings
     /// <summary>
-    /// Helper method to send the briefing text and sound to a list of entities
+    /// Gets an ordered player pool based on player preferences and the antagonist definition.
     /// </summary>
-    /// <param name="entities">The players chosen to be antags</param>
-    /// <param name="briefing">The briefing text to send</param>
-    /// <param name="briefingColor">The color the briefing should be, null for default</param>
-    /// <param name="briefingSound">The sound to briefing/greeting sound to play</param>
-    public void SendBriefing(List<EntityUid> entities, string briefing, Color? briefingColor, SoundSpecifier? briefingSound)
+    public AntagSelectionPlayerPool GetPlayerPool(Entity<AntagSelectionComponent> ent, List<ICommonSession> sessions, AntagSelectionDefinition def)
     {
-        foreach (var entity in entities)
+        var primaryList = new List<ICommonSession>();
+        var secondaryList = new List<ICommonSession>();
+        var fallbackList = new List<ICommonSession>();
+        var rawList = new List<ICommonSession>(sessions);
+        foreach (var session in sessions)
         {
-            SendBriefing(entity, briefing, briefingColor, briefingSound);
+            if (!IsSessionValid(ent, session, def))
+                continue;
+
+            if (!IsEntityValid(session.AttachedEntity, def))
+                continue;
+
+            var pref = (HumanoidCharacterProfile) _pref.GetPreferences(session.UserId).SelectedCharacter;
+            if (def.PrefRoles.Count == 0 || pref.AntagPreferences.Any(p => def.PrefRoles.Contains(p)))
+            {
+                primaryList.Add(session);
+            }
+            else if (def.PrefRoles.Count == 0 || pref.AntagPreferences.Any(p => def.FallbackRoles.Contains(p)))
+            {
+                secondaryList.Add(session);
+            }
+            else
+            {
+                fallbackList.Add(session);
+            }
+
+            rawList.Remove(session);
         }
```

Comment:
As above.

Also would be better to use RemoveAt instead of Remove, easiest thing is just iterate in reverse as otherwise it has to re-scan the entire list trying to find the session for every single one.